mri/reconstruct/linear.py

```diff
@@ -114,13 +116,185 @@ def l2norm(self, shape):
+
+
+class WaveletUD2(object):
+    """The wavelet undecimated operator using pysap wrapper.
+    """
+    def __init__(self, wavelet_id=24, nb_scale=4,
+                 multichannel=False, n_cpu=1, verbose=0):
+        """Init function for Undecimated wavelet transform
+
+        Parameters
+        -----------
+        wavelet_id: int, default 24 = undecimated (bi-) orthogonal transform
+            ID of wavelet being used
+        nb_scale: int, default 4
+            the number of scales in the decomposition.
+        multichannel: bool, default False
+            Boolean value to indicate if the incoming data is from
+            multiple-channels
+        n_cpu: int, default 0
+            Number of CPUs to run on. Only applicable if multichannel=True.
+        Private Variables:
+            _has_run: Checks if the get_mr_filters was called already
+        """
+        self.wavelet_id = wavelet_id
+        self.multichannel = multichannel
+        self.nb_scale = nb_scale
+        self.n_cpu = n_cpu
+        self._opt = [
+            '-t{}'.format(self.wavelet_id),
+            '-n{}'.format(self.nb_scale),
+        ]
+        self._has_run = False
+        self.coeffs_shape = None
+
+    def _get_filters(self, shape):
+        """Function to get the Wavelet coefficients of Delta[0][0].
+        This function is called only once and later the
+        wavelet coefficients are obtained by convolving these coefficients
+        with input Data
+        """
+        self.transform = get_mr_filters(
+            tuple(shape),
+            opt=self._opt,
+            coarse=True,
+        )
+        self._has_run = True
+
+    def _op(self, data):
+        """ Define the wavelet operator for single channel.
+        This is internal function that returns wavelet coefficients for a
+        single channel
+        Parameters
+        ----------
+        data: ndarray or Image
+            input 2D data array.
+
+        Returns
+        -------
+        coeffs: ndarray
+            the wavelet coefficients.
+        """
+        coefs_real = filter_convolve(data.real, self.transform)
+        coefs_imag = filter_convolve(data.imag, self.transform)
+        coeffs, coeffs_shape = flatten(
+            coefs_real + 1j * coefs_imag)
+        return coeffs, coeffs_shape
+
+    def op(self, data):
+        """ Define the wavelet operator.
+
+        This method returns the input data convolved with the wavelet filter.
+
+        Parameters
+        ----------
+        data: ndarray or Image
+            input 2D data array.
+
+        Returns
+        -------
+        coeffs: ndarray
+            the wavelet coefficients.
+        """
+        if not self._has_run:
+            if self.multichannel:
+                self._get_filters(list(data.shape)[1:])
+            else:
+                self._get_filters(data.shape)
+        if self.multichannel:
+            coeffs, self.coeffs_shape = zip(*Parallel(n_jobs=self.n_cpu)(
```

Comment:
Have you tried to see which backend was the fastest?
It might be interesting to leave that option to the user, given that it's not trivial determining which will be depending on the size of the data and the number of channels.
Look at this SO question for more on this: https://stackoverflow.com/questions/55956447/how-to-parallelize-scipy-fftconvolve-using-joblib

Reply:
Hmm, interesting, let me generalize this